File: examples/dekita/.claude/hooks/session-handoff-reader.py
#!/usr/bin/env python3
"""セッション開始時に前回の引き継ぎメモを読み込み表示。

Why:
    Claude Codeはセッション間で記憶を保持しない。前回の作業状態、
    未対応タスク、教訓を引き継ぐことで、継続性を確保する。

What:
    - セッション開始時（SessionStart）に発火
    - .claude/handoff/配下の有効なメモを読み込み
    - 自セッションと他セッションのメモを区別して表示
    - Git状態、オープンPR、ロック中worktreeも表示

State:
    - reads: .claude/handoff/*.json

Remarks:
    - 非ブロック型（情報表示のみ）
    - session-handoff-writerが生成、本フックが読み込み
    - メモの有効期間は24時間

Changelog:
    - silenvx/dekita#xxx: フック追加
    - silenvx/dekita#1333: 教訓抽出機能を追加
"""

from __future__ import annotations

import json
import sys
from datetime import UTC, datetime
from pathlib import Path
from typing import Any

# 共通モジュール
HOOKS_DIR = Path(__file__).parent
sys.path.insert(0, str(HOOKS_DIR))
from lib.execution import log_hook_execution
from lib.session import create_hook_context, parse_hook_input

# 引き継ぎメモの保存先
HANDOFF_DIR = HOOKS_DIR.parent / "handoff"

# 引き継ぎメモの有効期間（24時間以内の場合のみ表示）
HANDOFF_VALIDITY_HOURS = 24


def load_all_handoff_memos() -> list[dict[str, Any]]:
    """全ての有効なハンドオフメモを読み込み（時刻順）"""
    memos = []

    if not HANDOFF_DIR.exists():
        return memos

    for handoff_file in HANDOFF_DIR.glob("*.json"):
        try:
            with open(handoff_file, encoding="utf-8") as f:
                memo = json.load(f)
                if is_memo_valid(memo):
                    memos.append(memo)
        except (OSError, json.JSONDecodeError):
            continue

    # 生成時刻で降順ソート（最新が先頭）
    memos.sort(key=lambda m: m.get("generated_at", ""), reverse=True)

    return memos


def is_memo_valid(memo: dict[str, Any]) -> bool:
    """引き継ぎメモが有効期間内かチェック"""
    generated_at = memo.get("generated_at")
    if not generated_at:
        return False

    try:
        generated_time = datetime.fromisoformat(generated_at)
        now = datetime.now(UTC)
        age_hours = (now - generated_time).total_seconds() / 3600
        return age_hours < HANDOFF_VALIDITY_HOURS
    except (ValueError, TypeError):
        return False


def format_age(generated_at: str) -> str:
    """生成時刻から経過時間を文字列化"""
    try:
        generated_time = datetime.fromisoformat(generated_at)
        age_minutes = int((datetime.now(UTC) - generated_time).total_seconds() / 60)
        if age_minutes < 1:
            return "たった今"
        elif age_minutes < 60:
            return f"{age_minutes}分前"
        else:
            return f"{age_minutes // 60}時間前"
    except (ValueError, TypeError):
        return "不明"


def format_handoff_message(memos: list[dict[str, Any]], current_session_id: str) -> str:
    """複数のハンドオフメモを読みやすい形式にフォーマット"""
    if not memos:
        return ""

    lines = ["📝 **セッション引き継ぎ情報**", ""]

    # 自分のセッションのメモを優先して表示
    own_session_memos = [m for m in memos if m.get("session_id") == current_session_id]
    other_session_memos = [m for m in memos if m.get("session_id") != current_session_id]

    # 自分のセッションのメモがあればそれを使う、なければ最新のメモ
    if own_session_memos:
        latest = own_session_memos[0]
        is_own_session = True
    else:
        latest = memos[0]
        is_own_session = False

    session_label = "前回のセッション" if is_own_session else "別セッション"

    lines.append(
        f"**{session_label}からの引き継ぎ** ({format_age(latest.get('generated_at', ''))})"
    )
    lines.append("")

    # 作業状態
    work_status = latest.get("work_status", "不明")
    lines.append(f"**状態**: {work_status}")

    # 次のアクション
    next_action = latest.get("next_action", "")
    if next_action:
        lines.append(f"**次にすべきこと**: {next_action}")

    # 未対応タスク（あれば）
    pending_tasks = latest.get("pending_tasks", [])
    if pending_tasks:
        lines.append("")
        lines.append("**⚠️ 未対応タスク**:")
        for task in pending_tasks[:5]:
            lines.append(f"  - {task}")

    # 教訓・学び（あれば）
    lessons = latest.get("lessons_learned", [])
    if lessons:
        lines.append("")
        lines.append("**💡 前回の教訓**:")
        for lesson in lessons[:3]:
            lines.append(f"  - {lesson}")

    lines.append("")

    # Git状態
    git = latest.get("git", {})
    if git:
        branch = git.get("branch", "不明")
        uncommitted = git.get("uncommitted_changes", 0)
        untracked = git.get("untracked_files", 0)

        lines.append("**Git状態**:")
        lines.append(f"  - ブランチ: `{branch}`")
        if uncommitted > 0:
            lines.append(f"  - 未コミットの変更: {uncommitted}件 ⚠️")
        if untracked > 0:
            lines.append(f"  - 未追跡ファイル: {untracked}件")

    # オープンPR
    open_prs = latest.get("open_prs", [])
    if open_prs:
        lines.append("")
        lines.append("**オープンPR**:")
        for pr in open_prs[:3]:
            lines.append(
                f"  - #{pr.get('number')}: {pr.get('title', '')} (`{pr.get('branch', '')}`)"
            )

    # アクティブworktree
    worktrees = latest.get("worktrees", [])
    active_worktrees = [wt for wt in worktrees if wt.get("locked")]
    if active_worktrees:
        lines.append("")
        lines.append("**ロック中のworktree** (別セッションが作業中かも):")
        for wt in active_worktrees[:3]:
            lines.append(f"  - `{wt.get('branch', '?')}` @ {wt.get('path', '?')}")

    # セッションサマリー
    summary = latest.get("session_summary", {})
    if summary.get("blocks", 0) > 0:
        lines.append("")
        lines.append(f"**前回のセッション**: {summary.get('blocks', 0)}回ブロックされました")
        block_reasons = summary.get("block_reasons", [])
        if block_reasons:
            lines.append("  最近のブロック理由:")
            for reason in block_reasons[:2]:
                truncated = reason[:60]
                suffix = "..." if len(reason) > 60 else ""
                lines.append(f"    - {truncated}{suffix}")

    # 他セッションからのメモがある場合
    if other_session_memos:
        lines.append("")
        lines.append("---")
        lines.append(f"_他に{len(other_session_memos)}件の並列セッションの引き継ぎがあります_")

        # 重要なタスクや教訓があれば表示
        for memo in other_session_memos[:2]:
            pending = memo.get("pending_tasks", [])
            lessons = memo.get("lessons_learned", [])
            if pending or lessons:
                age = format_age(memo.get("generated_at", ""))
                lines.append(f"  ({age}):")
                for task in pending[:2]:
                    lines.append(f"    - ⚠️ {task}")
                for lesson in lessons[:1]:
                    lines.append(f"    - 💡 {lesson}")

    return "\n".join(lines)


def main():
    """SessionStart hookのエントリーポイント"""
    result = {"continue": True}

    try:
        # 現在のセッションIDを取得
        hook_input = parse_hook_input()

        ctx = create_hook_context(hook_input)
        current_session_id = hook_input.get("session_id") or ctx.get_session_id()

        # 全ての有効なメモを読み込み
        memos = load_all_handoff_memos()

        if memos:
            message = format_handoff_message(memos, current_session_id)
            if message:
                result["message"] = message

            log_hook_execution(
                "session-handoff-reader",
                "approve",
                "Handoff memos displayed",
                {
                    "memo_count": len(memos),
                    "latest_work_status": memos[0].get("work_status") if memos else None,
                    "has_pending_tasks": any(m.get("pending_tasks") for m in memos),
                    "has_lessons": any(m.get("lessons_learned") for m in memos),
                },
            )
        else:
            log_hook_execution("session-handoff-reader", "approve", "No valid handoff memos found")

    except Exception as e:
        # エラーがあっても継続
        log_hook_execution("session-handoff-reader", "approve", f"Error loading handoff memos: {e}")

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/session_handoff_reader.ts
#!/usr/bin/env bun
/**
 * セッション開始時に前回の引き継ぎメモを読み込み表示。
 *
 * Why:
 *   Claude Codeはセッション間で記憶を保持しない。前回の作業状態、
 *   未対応タスク、教訓を引き継ぐことで、継続性を確保する。
 *
 * What:
 *   - セッション開始時（SessionStart）に発火
 *   - .claude/handoff/配下の有効なメモを読み込み
 *   - 自セッションと他セッションのメモを区別して表示
 *   - Git状態、オープンPR、ロック中worktreeも表示
 *
 * State:
 *   - reads: .claude/handoff/*.json
 *
 * Remarks:
 *   - 非ブロック型（情報表示のみ）
 *   - session-handoff-writerが生成、本フックが読み込み
 *   - メモの有効期間は24時間
 *   - Python版: session_handoff_reader.py
 *
 * Changelog:
 *   - silenvx/dekita#xxx: フック追加
 *   - silenvx/dekita#1333: 教訓抽出機能を追加
 *   - silenvx/dekita#2917: TypeScript版初期実装
 *   - silenvx/dekita#3053: formatAgeの将来日時ハンドリング改善
 */

import { existsSync, readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { formatError } from "../lib/format_error";
import { logHookExecution } from "../lib/logging";
import { createContext, getSessionId, parseHookInput } from "../lib/session";

const HOOK_NAME = "session-handoff-reader";

// 引き継ぎメモの有効期間（24時間以内の場合のみ表示）
export const HANDOFF_VALIDITY_HOURS = 24;

interface GitInfo {
  branch?: string;
  uncommitted_changes?: number;
  untracked_files?: number;
}

interface PRInfo {
  number?: number;
  title?: string;
  branch?: string;
}

interface WorktreeInfo {
  path?: string;
  branch?: string;
  locked?: boolean;
}

interface SessionSummary {
  blocks?: number;
  block_reasons?: string[];
}

export interface HandoffMemo {
  session_id?: string;
  generated_at?: string;
  work_status?: string;
  next_action?: string;
  pending_tasks?: string[];
  lessons_learned?: string[];
  git?: GitInfo;
  open_prs?: PRInfo[];
  worktrees?: WorktreeInfo[];
  session_summary?: SessionSummary;
}

/**
 * Get project directory.
 */
function getProjectDir(): string {
  return process.env.CLAUDE_PROJECT_DIR || process.cwd();
}

/**
 * Get handoff directory.
 */
function getHandoffDir(): string {
  return join(getProjectDir(), ".claude", "handoff");
}

/**
 * Check if memo is within validity period.
 */
export function isMemoValid(memo: HandoffMemo): boolean {
  const generatedAt = memo.generated_at;
  if (!generatedAt) {
    return false;
  }

  try {
    const generatedTime = new Date(generatedAt).getTime();
    const now = Date.now();
    const ageHours = (now - generatedTime) / 1000 / 3600;
    return ageHours < HANDOFF_VALIDITY_HOURS;
  } catch {
    return false;
  }
}

/**
 * Load all valid handoff memos.
 */
function loadAllHandoffMemos(): HandoffMemo[] {
  const memos: HandoffMemo[] = [];
  const handoffDir = getHandoffDir();

  if (!existsSync(handoffDir)) {
    return memos;
  }

  try {
    const files = readdirSync(handoffDir);
    for (const file of files) {
      if (!file.endsWith(".json")) {
        continue;
      }

      try {
        const filePath = join(handoffDir, file);
        const content = readFileSync(filePath, "utf-8");
        const memo = JSON.parse(content) as HandoffMemo;
        if (isMemoValid(memo)) {
          memos.push(memo);
        }
      } catch {
        // 無効なJSONファイル、スキップ
      }
    }
  } catch {
    return memos;
  }

  // Sort by generated_at descending (newest first)
  memos.sort((a, b) => {
    const aTime = a.generated_at ? new Date(a.generated_at).getTime() : 0;
    const bTime = b.generated_at ? new Date(b.generated_at).getTime() : 0;
    return bTime - aTime;
  });

  return memos;
}

/**
 * Format age from generated_at timestamp.
 */
export function formatAge(generatedAt: string): string {
  try {
    if (!generatedAt) {
      return "不明";
    }
    const generatedTime = new Date(generatedAt).getTime();
    // Invalid Date returns NaN for getTime()
    if (Number.isNaN(generatedTime)) {
      return "不明";
    }
    const now = Date.now();
    const diffMs = now - generatedTime;

    // Handle future dates (clock skew, etc.)
    if (diffMs < 0) {
      return "不明";
    }

    const ageMinutes = Math.floor(diffMs / 1000 / 60);

    if (ageMinutes < 1) {
      return "たった今";
    }
    if (ageMinutes < 60) {
      return `${ageMinutes}分前`;
    }
    return `${Math.floor(ageMinutes / 60)}時間前`;
  } catch {
    return "不明";
  }
}

/**
 * Format handoff message from memos.
 */
export function formatHandoffMessage(
  memos: HandoffMemo[],
  currentSessionId: string | null,
): string {
  if (memos.length === 0) {
    return "";
  }

  const lines: string[] = ["📝 **セッション引き継ぎ情報**", ""];

  // Separate own session memos and other session memos
  const ownSessionMemos = memos.filter((m) => m.session_id === currentSessionId);
  const otherSessionMemos = memos.filter((m) => m.session_id !== currentSessionId);

  // Use own session memo if available, otherwise use the latest
  let latest: HandoffMemo;
  let isOwnSession: boolean;

  if (ownSessionMemos.length > 0) {
    latest = ownSessionMemos[0];
    isOwnSession = true;
  } else {
    latest = memos[0];
    isOwnSession = false;
  }

  const sessionLabel = isOwnSession ? "前回のセッション" : "別セッション";

  lines.push(`**${sessionLabel}からの引き継ぎ** (${formatAge(latest.generated_at || "")})`);
  lines.push("");

  // Work status
  const workStatus = latest.work_status || "不明";
  lines.push(`**状態**: ${workStatus}`);

  // Next action
  const nextAction = latest.next_action;
  if (nextAction) {
    lines.push(`**次にすべきこと**: ${nextAction}`);
  }

  // Pending tasks
  const pendingTasks = latest.pending_tasks || [];
  if (pendingTasks.length > 0) {
    lines.push("");
    lines.push("**⚠️ 未対応タスク**:");
    for (const task of pendingTasks.slice(0, 5)) {
      lines.push(`  - ${task}`);
    }
  }

  // Lessons learned
  const lessons = latest.lessons_learned || [];
  if (lessons.length > 0) {
    lines.push("");
    lines.push("**💡 前回の教訓**:");
    for (const lesson of lessons.slice(0, 3)) {
      lines.push(`  - ${lesson}`);
    }
  }

  lines.push("");

  // Git status
  const git = latest.git;
  if (git) {
    const branch = git.branch || "不明";
    const uncommitted = git.uncommitted_changes || 0;
    const untracked = git.untracked_files || 0;

    lines.push("**Git状態**:");
    lines.push(`  - ブランチ: \`${branch}\``);
    if (uncommitted > 0) {
      lines.push(`  - 未コミットの変更: ${uncommitted}件 ⚠️`);
    }
    if (untracked > 0) {
      lines.push(`  - 未追跡ファイル: ${untracked}件`);
    }
  }

  // Open PRs
  const openPrs = latest.open_prs || [];
  if (openPrs.length > 0) {
    lines.push("");
    lines.push("**オープンPR**:");
    for (const pr of openPrs.slice(0, 3)) {
      lines.push(`  - #${pr.number}: ${pr.title || ""} (\`${pr.branch || ""}\`)`);
    }
  }

  // Active worktrees
  const worktrees = latest.worktrees || [];
  const activeWorktrees = worktrees.filter((wt) => wt.locked);
  if (activeWorktrees.length > 0) {
    lines.push("");
    lines.push("**ロック中のworktree** (別セッションが作業中かも):");
    for (const wt of activeWorktrees.slice(0, 3)) {
      lines.push(`  - \`${wt.branch || "?"}\` @ ${wt.path || "?"}`);
    }
  }

  // Session summary
  const summary = latest.session_summary;
  if (summary && (summary.blocks || 0) > 0) {
    lines.push("");
    lines.push(`**前回のセッション**: ${summary.blocks}回ブロックされました`);
    const blockReasons = summary.block_reasons || [];
    if (blockReasons.length > 0) {
      lines.push("  最近のブロック理由:");
      for (const reason of blockReasons.slice(0, 2)) {
        const truncated = reason.slice(0, 60);
        const suffix = reason.length > 60 ? "..." : "";
        lines.push(`    - ${truncated}${suffix}`);
      }
    }
  }

  // Other session memos
  if (otherSessionMemos.length > 0) {
    lines.push("");
    lines.push("---");
    lines.push(`_他に${otherSessionMemos.length}件の並列セッションの引き継ぎがあります_`);

    // Show important tasks or lessons from other sessions
    for (const memo of otherSessionMemos.slice(0, 2)) {
      const pending = memo.pending_tasks || [];
      const memoLessons = memo.lessons_learned || [];
      if (pending.length > 0 || memoLessons.length > 0) {
        const age = formatAge(memo.generated_at || "");
        lines.push(`  (${age}):`);
        for (const task of pending.slice(0, 2)) {
          lines.push(`    - ⚠️ ${task}`);
        }
        for (const lesson of memoLessons.slice(0, 1)) {
          lines.push(`    - 💡 ${lesson}`);
        }
      }
    }
  }

  return lines.join("\n");
}

async function main(): Promise<void> {
  const result: { continue: boolean; message?: string } = { continue: true };
  let sessionId: string | undefined;

  try {
    const inputData = await parseHookInput();
    sessionId = inputData.session_id;
    const ctx = createContext(inputData);
    const currentSessionId = (inputData.session_id as string) || getSessionId(ctx);

    // Load all valid memos
    const memos = loadAllHandoffMemos();

    if (memos.length > 0) {
      const message = formatHandoffMessage(memos, currentSessionId);
      if (message) {
        result.message = message;
      }

      await logHookExecution(
        HOOK_NAME,
        "approve",
        "Handoff memos displayed",
        {
          memo_count: memos.length,
          latest_work_status: memos[0].work_status,
          has_pending_tasks: memos.some((m) => m.pending_tasks && m.pending_tasks.length > 0),
          has_lessons: memos.some((m) => m.lessons_learned && m.lessons_learned.length > 0),
        },
        { sessionId },
      );
    } else {
      await logHookExecution(HOOK_NAME, "approve", "No valid handoff memos found", undefined, {
        sessionId,
      });
    }
  } catch (error) {
    // Continue even on error
    await logHookExecution(
      HOOK_NAME,
      "approve",
      `Error loading handoff memos: ${formatError(error)}`,
      undefined,
      { sessionId },
    );
  }

  console.log(JSON.stringify(result));
}

// Only run main when executed directly, not when imported
if (import.meta.main) {
  main();
}
